### PR TITLE
ITS Algorithm 

### PR DESCRIPTION
**Pull request by aibobrov**

### CODE REVIEW

File: src/main/kotlin/org/jetbrains/research/ml/coding/assistant/system/finder/VertexFinders.kt [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+/**
+ * Finds the closest vertex to a given student's partial solution
+ */
+interface VertexFinder {
+    val matcher: PartialSolutionMatcher
+
+    fun findCorrespondingVertex(solutionSpace: SolutionSpace, partialSolution: PartialSolution): SolutionSpaceVertex?
+}
+
+class NaiveVertexFinder(override val matcher: BooleanPartialSolutionMatcher) : VertexFinder {
```
Comment (nbirillo) [inline src/main/kotlin/org/jetbrains/research/ml/coding/assistant/system/finder/VertexFinders.kt]: Please, add a detailed description

File: src/main/kotlin/org/jetbrains/research/ml/coding/assistant/system/finder/VertexFinders.kt [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    override fun findCorrespondingVertex(
+        solutionSpace: SolutionSpace,
+        partialSolution: PartialSolution
+    ): SolutionSpaceVertex? {
+        return solutionSpace.graph.vertexSet().firstOrNull { matcher.isMatched(it, partialSolution) }
+    }
+}
+
+class ParallelVertexFinder(override val matcher: PartialSolutionMatcher) : VertexFinder {
```
Comment (nbirillo) [inline src/main/kotlin/org/jetbrains/research/ml/coding/assistant/system/finder/VertexFinders.kt]: Please, add a detailed description

File: src/main/kotlin/org/jetbrains/research/ml/coding/assistant/system/matcher/PartialSolutionMatchers.kt [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+/**
+ * Matches the student's partial solution to the solution space's vertex
+ * Returns the difference score >= 0.0
+ * 0.0 means the vertex and the partial solution is perfectly matched.
+ */
+interface PartialSolutionMatcher {
+    fun differScore(vertex: SolutionSpaceVertex, partialSolution: PartialSolution): Double
+}
+
+interface BooleanPartialSolutionMatcher : PartialSolutionMatcher {
```
Comment (nbirillo) [inline src/main/kotlin/org/jetbrains/research/ml/coding/assistant/system/matcher/PartialSolutionMatchers.kt:20]: Please, add a detailed description

File: src/main/kotlin/org/jetbrains/research/ml/coding/assistant/system/matcher/PartialSolutionMatchers.kt [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    fun isMatched(vertex: SolutionSpaceVertex, partialSolution: PartialSolution): Boolean
+
+    override fun differScore(vertex: SolutionSpaceVertex, partialSolution: PartialSolution): Double {
+        return if (isMatched(vertex, partialSolution)) 0.0 else Double.POSITIVE_INFINITY
+    }
+}
+
+class ExactPartialSolutionMatcher : BooleanPartialSolutionMatcher {
```
Comment (nbirillo) [inline src/main/kotlin/org/jetbrains/research/ml/coding/assistant/system/matcher/PartialSolutionMatchers.kt:32]: Please, add a detailed description

File: src/main/kotlin/org/jetbrains/research/ml/coding/assistant/system/matcher/PartialSolutionMatchers.kt [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    override fun isMatched(vertex: SolutionSpaceVertex, partialSolution: PartialSolution): Boolean {
+        return vertex.fragment.root.isIsomorphicTo(partialSolution.context.root)
+    }
+}
+
+class TresholdSolutionMatcher(
```
Comment (nbirillo) [inline src/main/kotlin/org/jetbrains/research/ml/coding/assistant/system/matcher/PartialSolutionMatchers.kt]: Please, add a detailed description

File: src/main/kotlin/org/jetbrains/research/ml/coding/assistant/system/matcher/PartialSolutionMatchers.kt [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    private val threshold: Double,
+    private val matcher: PartialSolutionMatcher
+) : BooleanPartialSolutionMatcher {
+    override fun isMatched(vertex: SolutionSpaceVertex, partialSolution: PartialSolution): Boolean {
+        return matcher.differScore(vertex, partialSolution) < threshold
+    }
+}
+
+object EditPartialSolutionMatcher : PartialSolutionMatcher {
```
Comment (nbirillo) [inline src/main/kotlin/org/jetbrains/research/ml/coding/assistant/system/matcher/PartialSolutionMatchers.kt:65]: Please, add a detailed description

File: src/test/kotlin/org/jetbrains/research/ml/coding/assistant/system/finder/VertexFinderTest.kt [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+import org.jetbrains.research.ml.coding.assistant.unification.CompositeTransformation
+import org.jetbrains.research.ml.coding.assistant.unification.DatasetUnification
+import org.jetbrains.research.ml.coding.assistant.util.ParametrizedBaseWithSdkTest
+import org.junit.Ignore
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.junit.runners.Parameterized
+import java.io.File
+
+@Ignore
```
Comment (nbirillo) [inline src/test/kotlin/org/jetbrains/research/ml/coding/assistant/system/finder/VertexFinderTest.kt:22]: Why do you ignore this test?

File: src/test/kotlin/org/jetbrains/research/ml/coding/assistant/system/finder/VertexFinderTest.kt [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+@RunWith(Parameterized::class)
+class VertexFinderTest : ParametrizedBaseWithSdkTest(getResourcesRootPath(::VertexFinderTest)) {
+    @Test
+    fun testBasic() {
+        val inputDir = "/Users/artembobrov/Documents/masters/ast-transform/python/max_digit"
```
Comment (nbirillo) [inline src/test/kotlin/org/jetbrains/research/ml/coding/assistant/system/finder/VertexFinderTest.kt]: Please, delete absolute path

**Review by nbirillo — CHANGES_REQUESTED** [review]
Great!

File: src/test/kotlin/org/jetbrains/research/ml/coding/assistant/system/finder/VertexFinderTest.kt (hunk 7 above)
Comment (aibobrov) [reply src/test/kotlin/org/jetbrains/research/ml/coding/assistant/system/finder/VertexFinderTest.kt:22]: a little long to compute

**Review by elena-lyulina — APPROVED** [review]
Great, thanks

File: src/test/kotlin/org/jetbrains/research/ml/coding/assistant/system/finder/VertexFinderTest.kt (hunk 7 above)
Comment (elena-lyulina) [reply src/test/kotlin/org/jetbrains/research/ml/coding/assistant/system/finder/VertexFinderTest.kt:22]: Maybe it's better to create different tasks to run some tests locally so you can exclude some tests like this
```
test {
    exclude 'org/boo/**'
    exclude '**/Bar.class'
}
```
and on CI (so it doesn't matter how much time it takes to compute)

**Pull request merged**